kubelet: get IP based on service network IP mode for dual-stack support.

pmichali opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Allows kubelet to obtain the correct IP from pods, when operating in dual-stack mode.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #70653

**Special notes for your reviewer**:
Intent of this is to allow IPv4 only and IPv6 only to work as they do today, but adapt kubelet for when running in dual-stack cluster.

**Does this PR introduce a user-facing change?**:
NONE

/area ipv6
/sig network

pmichali commented:
Commit message for this change...

    This is WIP as it relies on bug fix 70633 that fixes reggression in etcd. Once
    that commit merges, this can be rebased.

    In GetPodIP, the kubelet code attempts to get the IPv4 address for the
    pod, and if that fails, tries to get the IPv6 address.  That works fine
    for IPv4 only and IPv6 only mode, but not for dual-stack, where each
    pod will have both addresses.

    In addition, since dual-stack will support only a single (selectable)
    family for services, we need to also ensure that kubelet is using the
    same family, when getting the IP. One way to do that is to request
    the IP, based on the family used for services.

    With the introduction of IPv6 only mode, a DNS_SVC_IP environment
    variable was defined, so that DNS used an IP that was in the family
    for services. The variable can be provided to the kubelet, via a
    drop-in file, as is done for kubeadm-dind-cluster.

    This change makes use of that information, reading the variable, and,
    if set, will use the same family as the DNS IP in the request to obtain
    the pod IP. Otherwise, it will fall back on trying to get the IPv4
    address, and then trying to get the IPv6 address, if no IPv4 address
    is available (to support backward compatibility).

pmichali commented:
NOTE: this is WIP because regression was found in etcd for IPv6 clusters and this commit includes the cherry picked fix under PR 70633. Once that is merged, a rebase can be done for this commit.

neolit123 commented on pkg/kubelet/dockershim/network/plugins.go:
```diff
@@ -247,18 +248,41 @@ func getOnePodIP(execer utilexec.Interface, nsenterPath, netnsPath, interfaceNam
-	if err != nil {
-		return nil, err
+	addrType := preferredFamily()
+        if addrType != "" {
```
small whitespace issue.

pmichali replied:
Will address when I rebase.

neolit123 replied:
this whitespace issue seems to be present still.

pmichali replied:
Sorry, I missed it. Had spaces vs tab, so it looked fine in my editor. Will be in next patch.

neolit123 commented on cmd/kubeadm/app/phases/etcd/local.go:
```diff
@@ -90,7 +92,7 @@ func CreateStackedEtcdStaticPodManifestFile(client clientset.Interface, manifest
 
 	// notifies the other members of the etcd cluster about the joining member
-	etcdPeerAddress := fmt.Sprintf("https://%s:%d", cfg.APIEndpoint.AdvertiseAddress, kubeadmconstants.EtcdListenPeerPort)
+	etcdPeerAddress := fmt.Sprintf("https://%s", net.JoinHostPort(cfg.APIEndpoint.AdvertiseAddress, strconv.Itoa(kubeadmconstants.EtcdListenPeerPort)))
```
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/etcd/local_test.go
would need some updates too?

neolit123 commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -282,7 +284,7 @@ func performEtcdStaticPodUpgrade(client clientset.Interface, waiter apiclient.Wa
 		return true, errors.Wrap(err, "failed to retrieve the current etcd version")
 	}
-	currentEtcdVersionStr, ok := currentEtcdVersions[fmt.Sprintf("https://%s:%d", cfg.APIEndpoint.AdvertiseAddress, constants.EtcdListenClientPort)]
+	currentEtcdVersionStr, ok := currentEtcdVersions[fmt.Sprintf("https://%s", net.JoinHostPort(cfg.APIEndpoint.AdvertiseAddress, strconv.Itoa(constants.EtcdListenClientPort)))]
```
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/staticpods_test.go
would need dual stack coverage too.

neolit123 reviewed (commented):
@pmichali thank you for working on this change.
added some minor comments.

this would need a release note instead of NONE.
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/priority important-longterm

Commit:
kubelet: get IP based on service network IP mode for dual-stack support.

In GetPodIP, the kubelet code attempts to get the IPv4 address for the
pod, and if that fails, tries to get the IPv6 address.  That works fine
for IPv4 only and IPv6 only mode, but not for dual-stack, where each
pod will have both addresses.

In addition, since dual-stack will support only a single (selectable)
family for services, we need to also ensure that kubelet is using the
same family, when getting the IP. One way to do that is to request
the IP, based on the family used for services.

With the previous introduction of IPv6 only mode, a DNS_SVC_IP environment
variable was defined, so that DNS used an IP that was in the family
for services. The variable can be provided to the kubelet, via a
drop-in file, as is done for kubeadm-dind-cluster.

This change makes use of that information, reading the variable, and,
if set, will use the same family as the DNS IP in the request to obtain
the pod IP. Otherwise, it will fall back on trying to get the IPv4
address, and then trying to get the IPv6 address, if no IPv4 address
is available (to support backward compatibility).

Fixes Issue: #70653

/area ipv6
/sig network

pmichali commented:
Updated commit, removing dependent changes from PR #70633, which has since merged. This is ready for review.

Two questions...

This uses the DNS_SVC_IP environment variable, which was introduced with IPv6 only capability previously, to determine if the IPv4 or IPv6 address is used for pod. If not specified, it will default back to trying to find the IPv4 address on the pod first, and then trying to find the IPv6 address.  Would the requirement of this environment variable constitute a "user facing" change?

If not, how do I remove the release note label? I accidentally deleted the text that is used to trigger this label, and only have the text "NONE", and I don't remember what the syntax is supposed to be.

pmichali commented:
/assign @dchen1107

pmichali commented:
Reviewers, PTAL, and let me know about the release note question I posed.  Thanks!

pmichali commented:
@neolit123 Can you take a look at the latest changes? This is ready for review. Thanks!

neolit123 commented on pkg/kubelet/dockershim/network/plugins.go:
```diff
@@ -247,18 +248,41 @@ func getOnePodIP(execer utilexec.Interface, nsenterPath, netnsPath, interfaceNam
+	glog.V(3).Infof("DNS Service IP is %s", dnsServiceIP)
+	dnsIP := net.ParseIP(dnsServiceIP)
+	if dnsIP == nil {
+		glog.Warningf("Unable to parse DNS_SVC_IP (%s) to determine preferred family", dnsServiceIP)
```
i think it make sense to indicate that the fall back here is to -4.

simplest way is to use a `goto` to a label before:
```
fallback:
	glog.V(3).Infof("Using IPv4.....
```
yet i see that `goto` is not that widely used in the project.

pmichali replied:
I'm changing the wording of the message. Let me know if that is OK.

neolit123 reviewed (approved):
@pmichali thanks for the update.
LGTM mostly. added a couple of minor comments.

sig-node and the kubelet maintainers would do the final LGTM / approve.
/ok-to-test

neolit123 commented:
also please add a release note as per the PR template:
https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md

neolit123 commented:
/remove-area kubeadm

pmichali reviewed (commented):
Will upload newer patch today.

pmichali commented:
@neolit123 Thanks for looking. I was wondering if a release note is needed for this change. See my post 5 days ago... I'm thinking not, but need guidance here.

pmichali commented:
@freehan @dchen1107 Can you PTAL at this kubelet change?  Also, regarding my question above, is a release note needed? If so, I could use a bit of guidance on the wording. If not, how do I remote the release note label?

pmichali commented:
@thockin  could you help getting some eyes on this. I'm running out of time to be able to work on getting this merged (taking on new position at work with limited upstream involvement). Thanks!

internetionals commented:
Still an issue on 1.14

/reopen
/remove-lifecycle-rotten
/remove-lifecycle-sta
